File: U1/EXAMEN/Comercializadora_Soap_DotNet_GR01/Comercializadora_Soap_DotNet_GR01/Services/FacturacionService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Comercializadora_Soap_DotNet_GR01.DataAcces.Repositories;
using Comercializadora_Soap_DotNet_GR01.DTOs;
using Comercializadora_Soap_DotNet_GR01.Models;

namespace Comercializadora_Soap_DotNet_GR01.Services
{
    public class FacturacionService
    {
        private readonly ProductoRepository _productoRepository;
        private readonly FacturaRepository _facturaRepository;

        public FacturacionService()
        {
            _productoRepository = new ProductoRepository();
            _facturaRepository = new FacturaRepository();
        }

        /// <summary>
        /// Genera una factura (EFECTIVO con 33% descuento o CREDITO sin descuento)
        /// </summary>
        /// <param name="solicitud">Solicitud con FormaPago: EFECTIVO o CREDITO.
        /// Si es CREDITO, NumeroCredito es requerido (obtenido desde BanQuito)</param>
        public FacturaDTO GenerarFactura(SolicitudFacturaDTO solicitud)
        {
            try
            {
                // Validaciones comunes
                if (solicitud == null)
                    throw new Exception("La solicitud de factura es nula");

                if (solicitud.Items == null || !solicitud.Items.Any())
                    throw new Exception("La solicitud de factura no contiene productos. Debe enviar al menos un producto.");

                if (string.IsNullOrWhiteSpace(solicitud.CedulaCliente))
                    throw new Exception("La cédula del cliente es requerida");

                if (string.IsNullOrWhiteSpace(solicitud.NombreCliente))
                    throw new Exception("El nombre del cliente es requerido");

                // Validar FormaPago
                if (string.IsNullOrWhiteSpace(solicitud.FormaPago))
                    throw new Exception("La forma de pago es requerida. Valores permitidos: EFECTIVO, CREDITO");

                string formaPago = solicitud.FormaPago.ToUpper();
                if (formaPago != "EFECTIVO" && formaPago != "CREDITO")
                    throw new Exception("Forma de pago inválida. Valores permitidos: EFECTIVO, CREDITO");

                // Validar NumeroCredito solo si es CREDITO
                if (formaPago == "CREDITO" && string.IsNullOrWhiteSpace(solicitud.NumeroCredito))
                    throw new Exception("El número de crédito es requerido para pagos a CREDITO. Debe obtenerlo desde el servicio de BanQuito.");

                // Crear factura
                var factura = new Factura
                {
                    NumeroFactura = _facturaRepository.GenerarNumeroFactura(),
                    CedulaCliente = solicitud.CedulaCliente,
                    NombreCliente = solicitud.NombreCliente,
                    FormaPago = formaPago,
                    NumeroCredito = formaPago == "CREDITO" ? solicitud.NumeroCredito : null,
                    FechaEmision = DateTime.Now
                };

                decimal subtotal = 0;
                var detalles = new List<DetalleFactura>();

                // Procesar cada item
                foreach (var item in solicitud.Items)
                {
                    var producto = _productoRepository.GetById(item.ProductoId);
                    if (producto == null)
                        throw new Exception($"Producto con ID {item.ProductoId} no encontrado");

                    if (!_productoRepository.TieneStock(item.ProductoId, item.Cantidad))
                        throw new Exception($"Stock insuficiente para el producto {producto.Nombre}");

                    var subtotalItem = producto.Precio * item.Cantidad;
                    subtotal += subtotalItem;

                    var detalle = new DetalleFactura
                    {
                        ProductoId = item.ProductoId,
                        Cantidad = item.Cantidad,
                        PrecioUnitario = producto.Precio,
                        Subtotal = subtotalItem
                    };

                    detalles.Add(detalle);

                    // Actualizar stock
                    _productoRepository.ActualizarStock(item.ProductoId, item.Cantidad);
                }

                // Calcular descuento según forma de pago
                factura.Subtotal = subtotal;
                if (formaPago == "EFECTIVO")
                {
                    // EFECTIVO: 33% de descuento
                    factura.Descuento = subtotal * 0.33m;
                    factura.Total = subtotal - factura.Descuento;
                }
                else
                {
                    // CREDITO: Sin descuento
                    factura.Descuento = 0;
                    factura.Total = subtotal;
                }

                factura.Detalles = detalles;

                // Guardar factura
                var facturaCreada = _facturaRepository.Create(factura);

                // Mapear a DTO
                return MapearFacturaADTO(facturaCreada);
            }
            catch (Exception ex)
            {
                throw new Exception($"Error al generar factura: {ex.Message}");
            }
        }

        /// <summary>
        /// Obtener facturas por cédula de cliente
        /// </summary>
        public List<FacturaDTO> ObtenerFacturasPorCliente(string cedula)
        {
            var facturas = _facturaRepository.GetByCedulaCliente(cedula);
            return facturas.Select(f => MapearFacturaADTO(f)).ToList();
        }

        /// <summary>
        /// Obtener factura por número
        /// </summary>
        public FacturaDTO ObtenerFacturaPorNumero(string numeroFactura)
        {
            var factura = _facturaRepository.GetByNumeroFactura(numeroFactura);
            return factura != null ? MapearFacturaADTO(factura) : null;
        }

        // Mapear Factura a DTO
        private FacturaDTO MapearFacturaADTO(Factura factura)
        {
            var facturaDto = new FacturaDTO
            {
                FacturaId = factura.FacturaId,
                NumeroFactura = factura.NumeroFactura,
                CedulaCliente = factura.CedulaCliente,
                NombreCliente = factura.NombreCliente,
                FormaPago = factura.FormaPago,
                Subtotal = factura.Subtotal,
                Descuento = factura.Descuento,
                Total = factura.Total,
                NumeroCredito = factura.NumeroCredito,
                FechaEmision = factura.FechaEmision,
                Detalles = new List<DetalleFacturaDTO>()
            };

            if (factura.Detalles != null)
            {
                foreach (var detalle in factura.Detalles)
                {
                    var producto = _productoRepository.GetById(detalle.ProductoId);
                    facturaDto.Detalles.Add(new DetalleFacturaDTO
                    {
                        ProductoId = detalle.ProductoId,
                        NombreProducto = producto?.Nombre ?? "Producto no encontrado",
                        Cantidad = detalle.Cantidad,
                        PrecioUnitario = detalle.PrecioUnitario,
                        Subtotal = detalle.Subtotal
                    });
                }
            }

            return facturaDto;
        }
    }
}

File: U1/PROYECTO/01 SERVIDOR/Comercializadora_Restful_Java_GR01/src/main/java/ec/edu/monster/service/FacturacionService.java
package ec.edu.monster.service;

import ec.edu.monster.dao.FacturaDAO;
import ec.edu.monster.dao.ProductoDAO;
import ec.edu.monster.dao.DetalleFacturaDAO;
import ec.edu.monster.dto.FacturaDTO;
import ec.edu.monster.dto.DetalleFacturaDTO;
import ec.edu.monster.dto.SolicitudCalculoDTO;
import ec.edu.monster.dto.CalculoFacturaDTO;
import ec.edu.monster.dto.DetalleCalculoDTO;
import ec.edu.monster.dto.ItemFacturaDTO;
import ec.edu.monster.model.Factura;
import ec.edu.monster.model.Producto;
import ec.edu.monster.model.DetalleFactura;
import jakarta.ejb.Stateless;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import java.math.BigDecimal;
import java.util.List;
import java.util.ArrayList;
import java.util.stream.Collectors;
import java.util.logging.Logger;

/**
 * Servicio para la lógica de negocio de Facturación
 * El sistema es independiente - para pagos a CREDITO, el cliente debe
 * obtener el numeroCredito directamente desde BanQuito antes de crear la factura
 */
@Stateless
public class FacturacionService {

    private static final Logger LOGGER = Logger.getLogger(FacturacionService.class.getName());

    @Inject
    private FacturaDAO facturaDAO;

    @Inject
    private ProductoDAO productoDAO;

    @Inject
    private DetalleFacturaDAO detalleFacturaDAO;
    
    /**
     * Convertir DetalleFactura a DTO
     */
    private DetalleFacturaDTO convertirDetalleADTO(DetalleFactura detalle) {
        DetalleFacturaDTO dto = new DetalleFacturaDTO();
        dto.setDetalleId(detalle.getDetalleId());
        dto.setProductoId(detalle.getProducto().getProductoId());
        dto.setCodigoProducto(detalle.getProducto().getCodigo());
        dto.setNombreProducto(detalle.getProducto().getNombre());
        dto.setCantidad(detalle.getCantidad());
        dto.setPrecioUnitario(detalle.getPrecioUnitario());
        dto.setSubtotal(detalle.getSubtotal());
        return dto;
    }
    
    /**
     * Convertir Factura a DTO
     */
    private FacturaDTO convertirFacturaADTO(Factura factura) {
        if (factura == null) return null;
        
        FacturaDTO dto = new FacturaDTO();
        dto.setFacturaId(factura.getFacturaId());
        dto.setNumeroFactura(factura.getNumeroFactura());
        dto.setCedulaCliente(factura.getCedulaCliente());
        dto.setNombreCliente(factura.getNombreCliente());
        dto.setFormaPago(factura.getFormaPago());
        dto.setSubtotal(factura.getSubtotal());
        dto.setDescuento(factura.getDescuento());
        dto.setTotal(factura.getTotal());
        dto.setNumeroCredito(factura.getNumeroCredito());
        dto.setFechaEmision(factura.getFechaEmision());
        
        // Convertir detalles
        if (factura.getDetalles() != null && !factura.getDetalles().isEmpty()) {
            List<DetalleFacturaDTO> detallesDTO = factura.getDetalles().stream()
                    .map(this::convertirDetalleADTO)
                    .collect(Collectors.toList());
            dto.setDetalles(detallesDTO);
        }
        
        return dto;
    }
    
    /**
     * Obtener todas las facturas
     */
    public List<FacturaDTO> obtenerTodas() {
        return facturaDAO.findAll().stream()
                .map(this::convertirFacturaADTO)
                .collect(Collectors.toList());
    }
    
    /**
     * Obtener factura por ID
     */
    public FacturaDTO obtenerPorId(Integer id) {
        Factura factura = facturaDAO.findByIdWithDetalles(id);
        return convertirFacturaADTO(factura);
    }
    
    /**
     * Obtener factura por número
     */
    public FacturaDTO obtenerPorNumero(String numeroFactura) {
        Factura factura = facturaDAO.findByNumeroFactura(numeroFactura);
        if (factura != null) {
            // Cargar detalles manualmente
            factura = facturaDAO.findByIdWithDetalles(factura.getFacturaId());
        }
        return convertirFacturaADTO(factura);
    }
    
    /**
     * Obtener facturas por cédula de cliente
     */
    public List<FacturaDTO> obtenerPorCliente(String cedula) {
        return facturaDAO.findByCedulaCliente(cedula).stream()
                .map(this::convertirFacturaADTO)
                .collect(Collectors.toList());
    }
    
    /**
     * Crear nueva factura
     *
     * FORMAS DE PAGO:
     * - EFECTIVO: Se aplica automáticamente un 33% de descuento sobre el subtotal
     * - CREDITO: Sin descuento. El cliente debe proporcionar un numeroCredito válido
     *            obtenido previamente desde el sistema BanQuito
     */
    @Transactional
    public FacturaDTO crearFactura(FacturaDTO facturaDTO) {
        // Validar que hay detalles
        if (facturaDTO.getDetalles() == null || facturaDTO.getDetalles().isEmpty()) {
            throw new RuntimeException("La factura debe tener al menos un detalle");
        }
        
        // Calcular totales primero
        BigDecimal subtotal = BigDecimal.ZERO;
        List<DetalleFactura> detalles = new ArrayList<>();
        
        // Procesar cada detalle y calcular subtotal
        for (DetalleFacturaDTO detalleDTO : facturaDTO.getDetalles()) {
            Producto producto = productoDAO.findById(detalleDTO.getProductoId());
            
            if (producto == null) {
                throw new RuntimeException("Producto no encontrado: " + detalleDTO.getProductoId());
            }
            
            if (producto.getStock() < detalleDTO.getCantidad()) {
                throw new RuntimeException("Stock insuficiente para producto: " + producto.getNombre());
            }
            
            BigDecimal subtotalDetalle = producto.getPrecio().multiply(new BigDecimal(detalleDTO.getCantidad()));
            subtotal = subtotal.add(subtotalDetalle);
        }

        // Calcular descuento automáticamente según forma de pago
        BigDecimal descuento;
        if ("EFECTIVO".equals(facturaDTO.getFormaPago())) {
            // EFECTIVO: 33% de descuento sobre el subtotal
            descuento = subtotal.multiply(new BigDecimal("0.33"));
            LOGGER.info("Pago en EFECTIVO - Descuento 33% aplicado: $" + descuento);
        } else if ("CREDITO".equals(facturaDTO.getFormaPago())) {
            // CREDITO: Sin descuento, pero requiere numeroCredito
            descuento = BigDecimal.ZERO;

            // Validar que el cliente proporcione el número de crédito
            if (facturaDTO.getNumeroCredito() == null || facturaDTO.getNumeroCredito().trim().isEmpty()) {
                throw new RuntimeException(
                    "El número de crédito es requerido para pagos a CREDITO. " +
                    "Debe obtenerlo desde el servicio de BanQuito antes de crear la factura."
                );
            }

            LOGGER.info("Pago a CREDITO - Sin descuento - Número de crédito: " + facturaDTO.getNumeroCredito());
        } else {
            throw new RuntimeException("Forma de pago no válida. Debe ser EFECTIVO o CREDITO");
        }

        BigDecimal total = subtotal.subtract(descuento);
        
        // Crear factura
        Factura factura = new Factura();
        factura.setNumeroFactura(facturaDAO.generarNumeroFactura());
        factura.setCedulaCliente(facturaDTO.getCedulaCliente());
        factura.setNombreCliente(facturaDTO.getNombreCliente());
        factura.setFormaPago(facturaDTO.getFormaPago());
        factura.setNumeroCredito(facturaDTO.getNumeroCredito()); // null para EFECTIVO, requerido para CREDITO
        factura.setSubtotal(subtotal);
        factura.setDescuento(descuento);
        factura.setTotal(total);
        
        // Crear detalles
        for (DetalleFacturaDTO detalleDTO : facturaDTO.getDetalles()) {
            Producto producto = productoDAO.findById(detalleDTO.getProductoId());
            
            // Crear detalle
            DetalleFactura detalle = new DetalleFactura();
            detalle.setFactura(factura);
            detalle.setProducto(producto);
            detalle.setCantidad(detalleDTO.getCantidad());
            detalle.setPrecioUnitario(producto.getPrecio());
            detalle.setSubtotal(producto.getPrecio().multiply(new BigDecimal(detalleDTO.getCantidad())));
            
            detalles.add(detalle);
            
            // Reducir stock
            productoDAO.reducirStock(producto.getProductoId(), detalleDTO.getCantidad());
        }
        
        factura.setDetalles(detalles);
        
        // Guardar factura (cascade guardará los detalles)
        Factura guardada = facturaDAO.save(factura);
        
        LOGGER.info("Factura creada exitosamente: " + guardada.getNumeroFactura());

        // Convertir factura guardada a DTO y retornar
        return convertirFacturaADTO(facturaDAO.findByIdWithDetalles(guardada.getFacturaId()));
    }
    
    /**
     * Obtener facturas a crédito
     */
    public List<FacturaDTO> obtenerFacturasCredito() {
        return facturaDAO.findFacturasCredito().stream()
                .map(this::convertirFacturaADTO)
                .collect(Collectors.toList());
    }
    
    /**
     * Buscar factura por número de crédito
     */
    public FacturaDTO obtenerPorNumeroCredito(String numeroCredito) {
        Factura factura = facturaDAO.findByNumeroCredito(numeroCredito);
        if (factura != null) {
            factura = facturaDAO.findByIdWithDetalles(factura.getFacturaId());
        }
        return convertirFacturaADTO(factura);
    }

    /**
     * Calcula el total de una factura SIN generarla
     * Útil para conocer el monto ANTES de solicitar crédito en BanQuito
     * NO actualiza stock, NO crea factura en BD
     *
     * @param solicitud Solicitud con lista de items (productoId y cantidad)
     * @return CalculoFacturaDTO con el total calculado y detalles
     */
    public CalculoFacturaDTO calcularTotalFactura(SolicitudCalculoDTO solicitud) {
        try {
            // Validaciones básicas
            if (solicitud == null) {
                throw new RuntimeException("La solicitud de factura es nula");
            }

            if (solicitud.getItems() == null || solicitud.getItems().isEmpty()) {
                throw new RuntimeException("La solicitud de factura no contiene productos. Debe enviar al menos un producto.");
            }

            BigDecimal subtotal = BigDecimal.ZERO;
            List<DetalleCalculoDTO> detalles = new ArrayList<>();

            // Calcular subtotal y validar productos
            for (ItemFacturaDTO item : solicitud.getItems()) {
                Producto producto = productoDAO.findById(item.getProductoId());

                if (producto == null) {
                    throw new RuntimeException("Producto con ID " + item.getProductoId() + " no encontrado");
                }

                if (producto.getStock() < item.getCantidad()) {
                    throw new RuntimeException(
                        String.format("Stock insuficiente para el producto '%s'. Stock disponible: %d, solicitado: %d",
                            producto.getNombre(), producto.getStock(), item.getCantidad())
                    );
                }

                BigDecimal subtotalItem = producto.getPrecio().multiply(new BigDecimal(item.getCantidad()));
                subtotal = subtotal.add(subtotalItem);

                DetalleCalculoDTO detalle = new DetalleCalculoDTO();
                detalle.setProductoId(item.getProductoId());
                detalle.setNombreProducto(producto.getNombre());
                detalle.setCantidad(item.getCantidad());
                detalle.setPrecioUnitario(producto.getPrecio());
                detalle.setSubtotal(subtotalItem);

                detalles.add(detalle);
            }

            // Retornar cálculo exitoso
            CalculoFacturaDTO resultado = new CalculoFacturaDTO();
            resultado.setExitoso(true);
            resultado.setMensaje("Cálculo realizado exitosamente");
            resultado.setTotal(subtotal);
            resultado.setDetalles(detalles);

            LOGGER.info("Cálculo de factura realizado. Total: $" + subtotal);

            return resultado;

        } catch (Exception ex) {
            // Retornar error con mensaje comprensible
            CalculoFacturaDTO resultado = new CalculoFacturaDTO();
            resultado.setExitoso(false);
            resultado.setMensaje(ex.getMessage());
            resultado.setTotal(BigDecimal.ZERO);
            resultado.setDetalles(new ArrayList<>());

            LOGGER.warning("Error en cálculo de factura: " + ex.getMessage());

            return resultado;
        }
    }
}
